tfjs-converter/src/executor/model_analysis.ts

```diff
@@ -156,21 +156,44 @@ export function getNodesInTopologicalOrder(
     }
   }
 
-  // Build a set for all nodes reachable by at least one predefined node.
-  // This can help us filter out redundant nodes from the returned node list.
-  // For example:
-  // If we have four nodes with dependencies like this:
-  //   a --> b --> c --> d
-  // when node `c` is predefined (e.g. given as an input tensor), we can
-  // skip node `a` and `b` since their outputs will never be used.
+  const orderedNodes = orderedNodeNames.map((name) => nameToNode.get(name));
+  const filteredOrderedNodes =
+      filterPredefinedReachableNodes(orderedNodes, predefinedNodes);
+
+  // TODO: Turn validation on/off with tf env flag.
+  validateNodesExecutionOrder(filteredOrderedNodes, predefinedNodes);
+
+  return filteredOrderedNodes;
+}
+
+/**
+ * This is a helper function of `getNodesInTopologicalOrder`.
+ * Returns ordered nodes reachable by at least one predefined node.
+ * This can help us filter out redundant nodes from the returned node list.
+ * For example:
+ * If we have four nodes with dependencies like this:
+ *   a --> b --> c --> d
+ * when node `c` is predefined (e.g. given as an input tensor), we can
+ * skip node `a` and `b` since their outputs will never be used.
+ *
+ * @param orderedNodes Graph nodes in execution order.
+ * @param predefinedNodes Graph inputs, weights, and init nodes. Nodes in this
+ *     list must have distinct names.
+ */
+function filterPredefinedReachableNodes(
+    orderedNodes: Node[], predefinedNodes: Node[]) {
+  const nameToNode =
+      new Map<string, Node>(orderedNodes.map((node) => [node.name, node]));
+
   // TODO: Filter out more nodes when >=2 nodes are predefined in a path.
   const stack = predefinedNodes.map((node) => node.name);
   const predefinedReachableNodeNames = new Set(stack);
   while (stack.length > 0) {
     const nodeName = stack.pop();
     const node = nameToNode.get(nodeName)!;
```

Comment:
Nit: Could add a short comment explaining what this does.
```suggestion
  // Perform a DFS starting from the set of all predefined nodes
  // to find the set of all nodes reachable from the predefined nodes.
  while (stack.length > 0) {
    const nodeName = stack.pop();
    const node = nameToNode.get(nodeName)!;
```